### ISSUE TITLE
Different output for custom tf.keras.Model within tf.function

### ISSUE THREAD

**Issue by hugoych** [tensorflow/tensorflow#44021]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow installed from (source or binary): Colab Pre-Installed
- TensorFlow version (use command below): 2.3.0
- Python version: 3.6.9
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source): 
- CUDA/cuDNN version: 
- GPU model and memory: 

**Describe the current behavior**
I'm trying to obtain reproducible results between graph and eager execution. While doing that I observed that calling my tf.keras.Model within a tf.function wouldn't exactly give me the same results as in eager execution. I was wondering if that's expected ? If yes, what's the reason for it ?

**Describe the expected behavior**
I would expect to get exactly the same value.

**Standalone code to reproduce the issue**
Here is a Colab : https://colab.research.google.com/drive/1n4p-Iq_g7dT-Vv5cb-s6nyeqjuYjyyGL?usp=sharing

**Other info / logs**

**Comment by Saduf2019** [comment]
I ran the code shared, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/77a8239694426f5c721cf17da53f4727/untitled435.ipynb).

**Comment by hugoych** [comment]
Hey,
Do you have any update on the matter @ymodak ? 

Thanks in advance!

**Comment by gideonite** [comment]
When I set `training=True` the issue goes away. When `training=False`about 90% of the values are the same between graph mode and eager mode.
```
np.isclose(encoder(inputs, False).numpy(), graph_call(inputs, False).numpy()).astype(np.int).mean()
=> 0.903798828125
```
I wouldn't be surprised if this is the expected behavior.

**Comment by fchollet** [comment]
Hi there,

The output in eager and graph appears to be the same -- the issue is with the code you use to check the difference.

```
a = encoder(inputs, False)
b = graph_call(inputs, False)
print(a)
print(b)
print(tf.reduce_max(a - b))  # prints 0
```

> When I set training=True the issue goes away

You are using `SyncBatchNormalization` in your model, which would introduce potential differences (hence triggering your check).

**Comment by hugoych** [comment]
> Hi there,
> 
> The output in eager and graph appears to be the same -- the issue is with the code you use to check the difference. Run the following to demonstrate that the outputs are identical:
> 
> ```
> a = encoder(inputs, False)
> b = graph_call(inputs, False)
> print(a)
> print(b)
> print(tf.reduce_max(a - b))  # prints 0
> ```
> 
> > When I set training=True the issue goes away
> 
> You are using `SyncBatchNormalization` in your model, which would introduce potential differences in training mode (hence triggering your check).

Hi,
Thanks for the answer! I'm sorry I had changed the colab without noticing. My issue was when `training=True`. 
You are right, this difference only occurs when using `tf.keras.layers.experimental.SyncBatchNormalization`. 

Do you have a idea why it is the case ? Especially here, we aren't using any distribution strategy so the layer should behave as a `tf.keras.layers.BatchNormalization` no ? The tensorflow documentation says "Without tf.distribute strategy, this layer behaves as a regular tf.keras.layers.BatchNormalization layer."

I modified the colab to highlight what I'm stating here.

**Comment by jvishnuvardhan** [comment]
@hugoych Difference in the results from eager and graph mode is very very low. I tried with the following tolerance levels and are close upto 6th decimal. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/64f870eb338876c070b621fcaed8c978/issue_tf_function.ipynb). 

```
tf.experimental.numpy.allclose(
    eager_true_s, graph_true_s, rtol=1e-06, atol=1e-06)

# ndarray<<tf.Tensor: shape=(), dtype=bool, numpy=True>>
```

Please close the issue if this was resolved for you. thanks!
